kube-client/src/api/remote_command.rs

```diff
@@ -202,71 +257,79 @@ where
+                        let status = serde_json::from_slice::<Status>(&bin[1..]).map_err(Error::DeserializeStatus)?;
                         status_tx.send(status).map_err(|_| Error::SendStatus)?;
-                        break;
-                    }
-
-                    Err(err) => {
+                        break
+                    },
+                    Some(Err(err)) => {
                         return Err(Error::ReceiveWebSocketMessage(err));
+                    },
+                    None => {
+                        // Connection closed properly
+                        break
+                    },
+                }
+            },
+            stdin_message = stdin_stream.next() => {
+                match stdin_message {
+                    Some(Ok(bytes)) => {
+                        if !bytes.is_empty() {
+                            let mut vec = Vec::with_capacity(bytes.len() + 1);
+                            vec.push(STDIN_CHANNEL);
+                            vec.extend_from_slice(&bytes[..]);
+                            server_send
+                                .send(ws::Message::binary(vec))
+                                .await
+                                .map_err(Error::SendStdin)?;
+                        }
+                    },
+                    Some(Err(err)) => {
+                        return Err(Error::ReadStdin(err));
+                    }
+                    None => {
+                        // Stdin closed (writer half dropped).
+                        // Let the server know and disconnect.
+                        server_send.close().await.map_err(Error::SendClose)?;
+                        break;
                     }
                 }
-                server_msg = server_recv.next();
-                next_stdin = p_next_stdin;
-            }
-
-            Left((None, _)) => {
-                // Connection closed properly
-                break;
-            }
-
-            // from stdin
-            Right((Some(Ok(bytes)), p_server_msg)) => {
-                if !bytes.is_empty() {
-                    let mut vec = Vec::with_capacity(bytes.len() + 1);
-                    vec.push(STDIN_CHANNEL);
-                    vec.extend_from_slice(&bytes[..]);
-                    server_send
-                        .send(ws::Message::binary(vec))
-                        .await
-                        .map_err(Error::SendStdin)?;
+            },
+            Some(terminal_size_message) = terminal_size_next, if have_terminal_size_rx => {
+                match terminal_size_message {
+                    Some(new_size) => {
+                        let mut vec = vec![RESIZE_CHANNEL];
+                        let mut serde_vec = serde_json::to_vec(&new_size).map_err(Error::SerializeTerminalSize)?;
+                        vec.append(&mut serde_vec);
+                        server_send.send(ws::Message::Binary(vec)).await.map_err(Error::SendTerminalSize)?;
```

Comment:
wait, this is a four byte u8 that we receive, parse as a length 4 vec of u8, and then sent  out?

this is different to the stdin handler which seems to use the channel name as a marker before appending the vector. is there any resources that explains what is going on here?

Reply:
Here I create a vec of u8 with one value in it `RESIZE_CHANNEL` that is equal to 4 to indicate the `resizeChannel` as show in [https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cri/streaming/remotecommand/websocket.go#L34](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cri/streaming/remotecommand/websocket.go#L34).
I'll add a comment to show from here theses constant come from.

Reply:
Thanks, it makes more sense now. I think I expected the `vec` to be created in a similar way to in the other parts of the loop ala:

```rust
let mut vec = Vec::with_capacity(new_size.len() + 1);
vec.push(RESIZE_CHANNEL);
vec.extend_from_slice(&new_size[..]);
```

but I guess the extra `serde_vec` is there to ensure the stream only actually passes `TerminalSize` objects around?